### ISSUE TITLE
custom upper limits on queue length?

### ISSUE THREAD

**Issue by agentzh** [kr/beanstalkd#56]
It'd be nice to have a size limit on queues specified by users. Do you have any plan to implement this?

It'll be very sad if the job producers can keep feeding jobs into a queue until the queue server eventually runs out of memory.

**Comment by JensRantil** [comment]
@agentzh Also, running out of memory doesn't crash the process, but simply makes producers unable to put new jobs until there is more available memory. See https://groups.google.com/d/msg/beanstalk-talk/O-AK3Hy_TrU/ABzGJo1XCwAJ.

The only use-case I can see for your feature request is if you have multiple tubes and you would like to make sure that producers for one tube doesn't get blocked by producers filling up the beanstalkd instance with jobs in another tube. I'd say this is a pretty rare case, and you are making things overly complicated if this is an issue.

@kr: I suggest this issue can be closed. @agentzh In case @kr closes this, feel free to reopen if still an issue!

**Comment by kr** [comment]
This is a reasonable feature request, but it has come up so rarely (and I'm not aware of anyone who's actually had a problem because of the absence of this feature) that it's probably not worth adding.

**Comment by agentzh** [comment]
@JensRantil For operating systems with memory over committing, running out of memory may make the OOM killer engage in I'm afraid.

@kr The lack of the queue length limit still looks like a severe limitation to me. And from the implementor's perspective, this should be dead simple to add IMHO.

**Comment by kr** [comment]
Yeah, I was probably too hasty in closing this bug. Sorry, I did a bunch on an airplane and didn't consider this carefully. I still agree it's a reasonable feature request. Just know that I have no idea if I'll have the time to implement it in the near future. I'd happily accept a patch for it.

Perhaps a new command:

```
limit <tube> <size>\r\n
```

that specifies either a number of jobs or a memory size in bytes (of the job payloads, or perhaps including the internal job structs as well). @agentzh which of these would be more useful?

Any tube with a specific limit would stick around. There would need to be a way to turn off/remove the limit, (e.g. `limit foo 0` or `limit foo unlimited` as special case).

Of if you have a better suggestion for the interface, that's great!

**Comment by JensRantil** [comment]
> Any tube with a specific limit would stick around. There would need to be a way to turn off/remove the limit, (e.g. limit foo 0 or limit foo unlimited as special case).

Maybe it would make more sense to add a command to [delete a tube](https://github.com/kr/beanstalkd/issues/25) instead. It would be slightly more general instead of `limit foo 0` etc. in case "persistent tubes" would be introduced by other settings such as limit.

**Comment by ysmolsky** [comment]
The spec of the feature is unclear. And there are not many voices asking the feature here.

**Comment by ysmolsky** [comment]
Protocol for existing commands needs to be changed for this feature. "put" command will return "tube is full" when the limit of the tube is reached.
